Migrate Python dependency From Python 2 to Python 3

<!--

Have you read Atom's Code of Conduct? By filing an Issue, you are expected to comply with it, including treating everyone with respect: https://github.com/atom/.github/blob/master/CODE_OF_CONDUCT.md

Do you want to ask a question? Are you looking for support? The Atom message board is the best place for getting support: https://discuss.atom.io

---

Keep in mind that Atom is highly customizable in a number of ways and we strongly prefer that you consider these options before filing this issue:

* https://flight-manual.atom.io/using-atom/sections/basic-customization/: tweak Atom's configuration, styles, and keybindings.
* https://flight-manual.atom.io/using-atom/sections/atom-packages/: install a community package.
* https://flight-manual.atom.io/hacking-atom/: use the Atom API in your init script, to create a package, or to enhance an existing package.

If you're convinced that none of these options are appropriate for the feature you want, please explain why that's the case by completely filling out the issue template below.

Also note that the Atom team has finite resources so it's unlikely that we'll work on feature requests. If we're interested in a particular feature however, we'll follow up and ask you to submit an RFC to talk about it in more detail.

-->

## Summary

Python 2 is being deprecated by Debian and Ubuntu, in favor of Python 3, in a move toward not shipping Python 2 at all. ([Python 2 is already End-Of-Life'd by the Python maintainers](https://www.python.org/psf/press-release/pr20191220/) [\*](http://pyfound.blogspot.com/2019/12/python-2-sunset.html?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+PythonSoftwareFoundationNews+%28Python+Software+Foundation+News%29)). In the long term, it will help Atom stay installable/working on these distros to specify a Python 3 dependency (in addition to or instead of Python 2). Even longer-term, Python 2 will no-longer be available, so it will make sense to drop it entirely. If Python 2 gives Atom any trouble in the medium-term, it can simply be dropped.

## Motivation

<!-- Why are we doing this? What use cases does it support? What is the expected outcome? -->

Doing this keeps Atom installable and working on newer distros/operating systems.

Furthermore, making sure the build process doesn't require Python 2 makes it easy to continue building Atom once Python 2 is End-of-Life'd.

## Describe alternatives you've considered

<!-- A clear and concise description of the alternative solutions you've considered. Be sure to explain why Atom's existing customizability isn't suitable for this feature. -->

- Un-bundle `node-gyp` and remove any other usage of Python from within this package.
- OR: Allow users to install Atom on all platforms without requiring Python on their system.

## Additional context

<!-- Add any other context or screenshots about the feature request here. -->

- According to [this comment](https://github.com/atom/atom/issues/20356#issuecomment-578379533), Python is only needed for node-gyp.
  - [Node-gyp currently works with both Python 2.7.x and Python 3.5.x or newer.](https://github.com/nodejs/node-gyp#on-unix)
- See #20406 for some prior discussion, see #20356 for lots more discussion. These are examples of the kinds of packaging troubles that might arise relating to the Python 2 EOL and transition to Python 3.
- I am personally more of a Debian/Ubuntu person myself, but this applies to all OSes and platforms.

<details><summary>Full error message (click to expand)</summary>

```
$ apm install tree-view
The tree-view package is bundled with Atom and should not be explicitly installed.
You can run `apm uninstall tree-view` to uninstall it and then the version bundled
with Atom will be used.
Installing tree-view to /home/[user]/.atom/packages ✗

> pathwatcher@8.1.0 install /tmp/apm-install-dir-12033-11366-ljo9wx.osrxt/node_modules/tree-view/node_modules/pathwatcher
> node-gyp rebuild


npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
gyp ERR! configure error 
gyp ERR! stack Error: Command failed: /usr/share/atom/resources/app/apm/bin/python-interceptor.sh -c import platform; print(platform.python_version());
gyp ERR! stack /usr/share/atom/resources/app/apm/bin/python-interceptor.sh: line 40: exec: python: not found
gyp ERR! stack 
gyp ERR! stack     at ChildProcess.exithandler (child_process.js:291:12)
gyp ERR! stack     at ChildProcess.emit (events.js:182:13)
gyp ERR! stack     at maybeClose (internal/child_process.js:961:16)
gyp ERR! stack     at Socket.stream.socket.on (internal/child_process.js:380:11)
gyp ERR! stack     at Socket.emit (events.js:182:13)
gyp ERR! stack     at Pipe._handle.close [as _onclose] (net.js:596:12)
gyp ERR! System Linux 5.4.0-21-generic
gyp ERR! command "/usr/share/atom/resources/app/apm/bin/node" "/usr/share/atom/resources/app/apm/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /tmp/apm-install-dir-12033-11366-ljo9wx.osrxt/node_modules/tree-view/node_modules/pathwatcher
gyp ERR! node -v v10.2.1
gyp ERR! node-gyp -v v3.7.0
gyp ERR! not ok 
npm WARN enoent ENOENT: no such file or directory, open '/tmp/apm-install-dir-12033-11366-ljo9wx.osrxt/package.json'
npm WARN apm-install-dir-12033-11366-ljo9wx.osrxt No description
npm WARN apm-install-dir-12033-11366-ljo9wx.osrxt No repository field.
npm WARN apm-install-dir-12033-11366-ljo9wx.osrxt No README data
npm WARN apm-install-dir-12033-11366-ljo9wx.osrxt No license field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pathwatcher@8.1.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the pathwatcher@8.1.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/[user]/.atom/.apm/_logs/2020-04-03T16_22_59_288Z-debug.log
```

</details>

<details><summary>`atom --version` output (click to expand)</summary>

```
$ atom --version
Atom    : 1.45.0
Electron: 4.2.7
Chrome  : 69.0.3497.128
Node    : 10.11.0
```

</details>

<details><summary>`apm --version` output (click to expand)</summary>

```
$ apm --version
apm  2.4.3
npm  6.2.0
node 10.2.1 x64
atom 1.45.0
python 3.8.2
git 2.25.1
```

</details>

Can confirm the Windows packaging doesn't include Python.

Zipped apm `apm -v`:

```
C:\Users\[User]\Downloads\atom-x64-windows\Atom x64\resources\cli>.\apm.cmd -v
apm  2.4.3
npm  6.2.0
node 10.2.1 x64
atom 1.45.0
python
git
visual studio
```

Error message when trying to `apm install tree-view` (no Python):
```
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
```

<details><summary>Full error message (click to expand):</summary>

```
C:\Users\[User]\Downloads\atom-x64-windows\Atom x64\resources\cli>.\apm.cmd install tree-view
The tree-view package is bundled with Atom and should not be explicitly installed.
You can run `apm uninstall tree-view` to uninstall it and then the version bundled
with Atom will be used.
Installing tree-view to C:\Users\[User]\.atom\packages failed

> pathwatcher@8.1.0 install C:\Users\[User]\AppData\Local\Temp\apm-install-dir-120318-4456-jo76ma.jg0c\node_modules\tree-view\node_modules\pathwatcher
> node-gyp rebuild


C:\Users\[User]\AppData\Local\Temp\apm-install-dir-120318-4456-jo76ma.jg0c\node_modules\tree-view\node_modules\pathwatcher>if not defined npm_config_node_gyp (node "C:\Users\[User]\Downloads\atom-x64-windows\Atom x64\resources\app\apm\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Users\[User]\Downloads\atom-x64-windows\Atom x64\resources\app\apm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )

npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (C:\Users\[User]\Downloads\atom-x64-windows\Atom x64\resources\app\apm\node_modules\npm\node_modules\node-gyp\lib\configure.js:492:19)
gyp ERR! stack     at PythonFinder.<anonymous> (C:\Users\[User]\Downloads\atom-x64-windows\Atom x64\resources\app\apm\node_modules\npm\node_modules\node-gyp\lib\configure.js:517:16)
gyp ERR! stack     at C:\Users\[User]\Downloads\atom-x64-windows\Atom x64\resources\app\apm\node_modules\npm\node_modules\graceful-fs\polyfills.js:284:29
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:182:21)
gyp ERR! System Windows_NT 10.0.18363
gyp ERR! command "C:\\Users\\[User]\\Downloads\\atom-x64-windows\\Atom x64\\resources\\app\\apm\\bin\\node.exe" "C:\\Users\\[User]\\Downloads\\atom-x64-windows\\Atom x64\\resources\\app\\apm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\[User]\AppData\Local\Temp\apm-install-dir-120318-4456-jo76ma.jg0c\node_modules\tree-view\node_modules\pathwatcher
gyp ERR! node -v v10.2.1
gyp ERR! node-gyp -v v3.7.0
gyp ERR! not ok
npm WARN enoent ENOENT: no such file or directory, open 'C:\Users\[User]\AppData\Local\Temp\apm-install-dir-120318-4456-jo76ma.jg0c\package.json'
npm WARN apm-install-dir-120318-4456-jo76ma.jg0c No description
npm WARN apm-install-dir-120318-4456-jo76ma.jg0c No repository field.
npm WARN apm-install-dir-120318-4456-jo76ma.jg0c No README data
npm WARN apm-install-dir-120318-4456-jo76ma.jg0c No license field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pathwatcher@8.1.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the pathwatcher@8.1.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\[User]\.atom\.apm\_logs\2020-04-18T17_22_03_188Z-debug.log
```

</details>

Installed apm `apm -v`:

```
C:\Users\[User]\AppData\Local\atom\app-1.45.0\resources\cli>.\apm.cmd -v
apm  2.4.3
npm  6.2.0
node 10.2.1 x64
atom 1.45.0
python
git
visual studio
```

Installed apm, error message when doing `apm install tree-view`:
```
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
```

<details><summary>Full error message (click to expand):</summary>

```
C:\Users\[User]\AppData\Local\atom\app-1.45.0\resources\cli>.\apm.cmd install tree-view
The tree-view package is bundled with Atom and should not be explicitly installed.
You can run `apm uninstall tree-view` to uninstall it and then the version bundled
with Atom will be used.
Installing tree-view to C:\Users\[User]\.atom\packages failed

> pathwatcher@8.1.0 install C:\Users\[User]\AppData\Local\Temp\apm-install-dir-120318-16440-ycddwg.rtoo\node_modules\tree-view\node_modules\pathwatcher
> node-gyp rebuild


C:\Users\[User]\AppData\Local\Temp\apm-install-dir-120318-16440-ycddwg.rtoo\node_modules\tree-view\node_modules\pathwatcher>if not defined npm_config_node_gyp (node "C:\Users\[User]\AppData\Local\atom\app-1.45.0\resources\app\apm\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Users\[User]\AppData\Local\atom\app-1.45.0\resources\app\apm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )

npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (C:\Users\[User]\AppData\Local\atom\app-1.45.0\resources\app\apm\node_modules\npm\node_modules\node-gyp\lib\configure.js:492:19)
gyp ERR! stack     at PythonFinder.<anonymous> (C:\Users\[User]\AppData\Local\atom\app-1.45.0\resources\app\apm\node_modules\npm\node_modules\node-gyp\lib\configure.js:517:16)
gyp ERR! stack     at C:\Users\[User]\AppData\Local\atom\app-1.45.0\resources\app\apm\node_modules\npm\node_modules\graceful-fs\polyfills.js:284:29
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:182:21)
gyp ERR! System Windows_NT 10.0.18363
gyp ERR! command "C:\\Users\\[User]\\AppData\\Local\\atom\\app-1.45.0\\resources\\app\\apm\\bin\\node.exe" "C:\\Users\\[User]\\AppData\\Local\\atom\\app-1.45.0\\resources\\app\\apm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\[User]\AppData\Local\Temp\apm-install-dir-120318-16440-ycddwg.rtoo\node_modules\tree-view\node_modules\pathwatcher
gyp ERR! node -v v10.2.1
gyp ERR! node-gyp -v v3.7.0
gyp ERR! not ok
npm WARN enoent ENOENT: no such file or directory, open 'C:\Users\[User]\AppData\Local\Temp\apm-install-dir-120318-16440-ycddwg.rtoo\package.json'
npm WARN apm-install-dir-120318-16440-ycddwg.rtoo No description
npm WARN apm-install-dir-120318-16440-ycddwg.rtoo No repository field.
npm WARN apm-install-dir-120318-16440-ycddwg.rtoo No README data
npm WARN apm-install-dir-120318-16440-ycddwg.rtoo No license field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pathwatcher@8.1.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the pathwatcher@8.1.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\[User]\.atom\.apm\_logs\2020-04-18T17_54_02_053Z-debug.log
```

</details>

Miscellaneous note: I also noticed a package that won't install without Git on the system.

```
Installing “ide-rust@0.21.2” failed. [Hide output…]
Failed to install ide-rust because Git was not found.

The ide-rust package has module dependencies that cannot be installed without Git.

You need to install Git and add it to your path environment variable in order to install this package.

You can install Git by downloading, installing, and launching GitHub for Windows: https://windows.github.com

Run apm -v after installing Git to see what version has been detected.
```

Final note: many packages can be successfully installed without using node-gyp (and thus Python) nor using Git, so there is some lee-way regarding if either of those should be bundled, or if they should be considered install-time dependencies at all. Maybe bundling or requiring Python on the system at install-time for Atom is heavy-handed?

--- a/resources/linux/debian/control.in
+++ b/resources/linux/debian/control.in
@@ -1,6 +1,6 @@
 Package: <%= appFileName %>
 Version: <%= version %>
-Depends: git, libgconf-2-4 (>= 3.2.5) | libgconf2-4, libgtk-3-0 (>= 3.9.10), libgcrypt11 | libgcrypt20, libnotify4, libxtst6, libnss3 (>= 2:3.22), python, gvfs-bin, xdg-utils, libx11-xcb1, libxss1, libasound2 (>= 1.0.16), libxkbfile1, libcurl3 | libcurl4, policykit-1
+Depends: git, libgconf-2-4 (>= 3.2.5) | libgconf2-4, libgtk-3-0 (>= 3.9.10), libgcrypt11 | libgcrypt20, libnotify4, libxtst6, libnss3 (>= 2:3.22), gvfs-bin, xdg-utils, libx11-xcb1, libxss1, libasound2 (>= 1.0.16), libxkbfile1, libcurl3 | libcurl4, policykit-1
 Recommends: lsb-release
 Suggests: libsecret-1-0, gir1.2-gnomekeyring-1.0
 Section: devel

--- a/resources/linux/debian/control.in
+++ b/resources/linux/debian/control.in
@@ -1,6 +1,6 @@
 Package: <%= appFileName %>
 Version: <%= version %>
-Depends: git, libgconf-2-4 (>= 3.2.5) | libgconf2-4, libgtk-3-0 (>= 3.9.10), libgcrypt11 | libgcrypt20, libnotify4, libxtst6, libnss3 (>= 2:3.22), python | python2, gvfs-bin, xdg-utils, libx11-xcb1, libxss1, libasound2 (>= 1.0.16), libxkbfile1, libcurl3 | libcurl4, policykit-1
+Depends: git, libgconf-2-4 (>= 3.2.5) | libgconf2-4, libgtk-3-0 (>= 3.9.10), libgcrypt11 | libgcrypt20, libnotify4, libxtst6, libnss3 (>= 2:3.22), python, gvfs-bin, xdg-utils, libx11-xcb1, libxss1, libasound2 (>= 1.0.16), libxkbfile1, libcurl3 | libcurl4, policykit-1
 Recommends: lsb-release
 Suggests: libsecret-1-0, gir1.2-gnomekeyring-1.0
 Section: devel

